cmd/kubeadm/app/node/bootstrap.go

```diff
@@ -0,0 +1,138 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/golang/glog"
+
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/api"
+	kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
+	"k8s.io/kubernetes/pkg/apis/certificates"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/client/typed/discovery"
+	"k8s.io/kubernetes/pkg/client/unversioned/clientcmd"
+	"k8s.io/kubernetes/pkg/types"
+	"k8s.io/kubernetes/pkg/util/wait"
+)
+
+// RetryTimeout between the subsequent attempts to connect
+// to an API endpoint
+const retryTimeout = 5
+
+// EstablishMasterConnection with exactly one of the provided API endpoints.
+// The method builds a client for every endpoint and concurrently
+// tries to connect to one of the provided endpoints every `retryTimeout` seconds.
+func EstablishMasterConnection(s *kubeadmapi.KubeadmConfig, clusterInfo *kubeadmapi.ClusterInfo) (*kubeadmapi.ConnectionDetails, error) {
+	hostName, err := os.Hostname()
+	if err != nil {
+		return nil, fmt.Errorf("<node/csr> failed to get node hostname [%v]", err)
+	}
+
+	// TODO: hostname == nodename doesn't hold on all clouds (AWS).
+	// But we don't have a cloudprovider, so we're stuck.
+	glog.Errorf("assuming that hostname is the same as NodeName")
+	nodeName := types.NodeName(hostName)
+
+	if err != nil {
+		return nil, fmt.Errorf("<node/bootstrap-client> failed to get hostname for node: [%v]", err)
+	}
+	endpoints := clusterInfo.Endpoints
+	caCerts := clusterInfo.CertificateAuthorities
+	if len(endpoints) != len(caCerts) {
+		return nil, fmt.Errorf("<node/bootstrap-client> number of provided API endpoints does not match " +
+			"the number of root certificates.")
+	}
+
+	stopChan := make(chan struct{})
+	result := make(chan *kubeadmapi.ConnectionDetails)
+
+	for i, endpoint := range endpoints {
+		caCert := []byte(caCerts[i])
+		clientSet, err := createClients(caCert, endpoint, s.Secrets.BearerToken, nodeName)
+		if err != nil {
+			glog.Warningf("<node/bootstrap-client> Warning: %s. Skipping endpoint %s.", err, endpoint)
+			continue
+		}
+		go wait.Until(func() {
+			fmt.Printf("<node/bootstrap-client> Trying to connect to endpoint %s\n", endpoint)
+			if err := checkCertsAPI(clientSet.DiscoveryClient); err != nil {
+				glog.Warningf("<node/bootstrap-client> Failed to connect to %s: %v", endpoint, err)
+				return
```

Comment:
Just one comment on this: shouldn't the `wg.Add(1)` be outside of the for loop? In my eyes the goal here is to connect to **at least one** of the API endpoints and bootstrap with it, not necessarily to all of them. WDYT?